apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,408 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_resource_manager).
+-behaviour(gen_statem).
+
+-include("emqx_resource.hrl").
+-include("emqx_resource_utils.hrl").
+
+% API
+-export([
+    create/5,
+    create_dry_run/2,
+    ets_lookup/1,
+    get_metrics/1,
+    health_check/1,
+    list_all/0,
+    list_group/1,
+    lookup/1,
+    recreate/4,
+    remove/1,
+    reset_metrics/1,
+    restart/2,
+    set_resource_status_connecting/1,
+    stop/1
+]).
+
+% Server
+-export([start_link/5]).
+
+% Behaviour
+-export([init/1, callback_mode/0, handle_event/4, terminate/3]).
+
+% State record
+-record(data, {id, group, mod, config, opts, status, state, error}).
+
+-define(SHORT_HEALTHCHECK_INTERVAL, 1000).
+-define(HEALTHCHECK_INTERVAL, 15000).
+-define(ETS_TABLE, emqx_resource_manager).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+%% @doc Called from emqx_resource when starting a resource instance.
+%%
+%% Triggers the emqx_resource_manager_sup supervisor to actually create
+%% and link the process itself if not already started.
+-spec create(
+    instance_id(),
+    resource_group(),
+    resource_type(),
+    resource_config(),
+    create_opts()
+) -> {ok, resource_data() | 'already_created'} | {error, Reason :: term()}.
+create(InstId, Group, ResourceType, Config, Opts) ->
+    case lookup(InstId) of
+        {ok, _, _} ->
+            {ok, already_created};
```

Comment:
Is it possible to make this function idempotent, so it always returns resource data in `ok` case? Maybe name it `ensure_resource`, because this is what it does.